packages/playwright/src/worker/testInfo.ts

```diff
@@ -311,18 +305,20 @@ export class TestInfoImpl implements TestInfo {
               break;
             }
           }
-          error = step.error;
         }
 
         const payload: StepEndPayload = {
           testId: this._test.id,
           stepId,
           wallTime: step.endWallTime,
-          error,
+          error: step.error,
         };
         this._onStepEnd(payload);
-        const errorForTrace = error ? { name: '', message: error.message || '', stack: error.stack } : undefined;
+        const errorForTrace = step.error ? { name: '', message: step.error.message || '', stack: step.error.stack } : undefined;
         this._tracing.appendAfterActionForStep(stepId, errorForTrace, result.attachments);
+
+        if (step.isSoft && result.error)
+          this._failWithError(result.error, false /* isHardError */);
```

Comment:
Instead of setting a symbol, perhaps we can just pass the step here directly? If that does not work, we need a comment that explains the situation where we have to pass the step through the symbol.